### PR TITLE
[blueprints] add VS Code schema-installing CLI command

### PR DESCRIPTION
**Pull request by benpankow**
## Summary

First pass implementation of a `dagster-blueprints configure-vscode` CLI command which installs a lightweight VS Code extention that provides JSON schemas to describe user-defined Blueprints.

Users can specify which module to read loaders from + which vscode folder to use as a staging are, or infer from the environment.

Modeled after #22017, but uses an extension in place of editing `settings.json` inplace to avoid adjusting user config + because patching commented JSON is tricky.

https://github.com/dagster-io/dagster/assets/10215173/ed8a4c0a-bbc0-4ff8-ad51-9e1e24a00645

## Test Plan

Tested locally, unit tests.

### CODE REVIEW

**Comment by benpankow** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22186?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#22186** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22186?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#22184** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22184?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @benpankow and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: examples/experimental/dagster-blueprints/dagster_blueprints/cli.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+from dagster._core.remote_representation.origin import ManagedGrpcPythonEnvCodeLocationOrigin
+from dagster._core.workspace.load_target import PyProjectFileTarget
+from dagster._utils.warnings import disable_dagster_warnings
+
+from dagster_blueprints.load_from_yaml import YamlBlueprintsLoader
+
+from .version import __version__
+
+
+def infer_vscode_path(base_path: Path) -> Optional[str]:
```
Comment (benpankow) [inline examples/experimental/dagster-blueprints/dagster_blueprints/cli.py:23]: Right now we store the built extension in the `.vscode` directory - maybe we should put it somewhere else?

File: examples/experimental/dagster-blueprints/dagster_blueprints/cli.py [hunk 2]
```diff
@@ -55,6 +70,49 @@ def generate_schemas(loader_module: Optional[str] = None):
...
+)
+@click.option(
+    "--loader-module",
+    type=click.STRING,
+    help="Path of Python module that contains YamlBlueprintsLoader objects. Defaults to Dagster project module, if `pyproject.toml` exists.",
+)
+@click.option(
+    "--vscode-folder-path",
+    type=click.STRING,
+    help="Path to the VS Code workspace folder. If not provided, attempts to find the .vscode directory in the current directory or any of its parents.",
```
Comment (sryza) [inline examples/experimental/dagster-blueprints/dagster_blueprints/cli.py:84]: Is "workspace folder" accurate VS Code terminology? I find the terminology a bit mind boggling: https://stackoverflow.com/questions/44629890/what-is-a-workspace-in-visual-studio-code.

File: examples/experimental/dagster-blueprints/dagster_blueprints/cli.py [hunk 3]
```diff
@@ -55,6 +70,49 @@ def generate_schemas(loader_module: Optional[str] = None):
...
+)
+def configure_vscode(loader_module: Optional[str] = None, vscode_folder_path: Optional[str] = None):
```
Comment (sryza) [inline examples/experimental/dagster-blueprints/dagster_blueprints/cli.py:86]:
```suggestion
def configure_vscode(loader_module: Optional[str] = None, vscode_folder_path: Optional[str] = None) -> None:
```

File: examples/experimental/dagster-blueprints/dagster_blueprints/cli.py [hunk 4]
```diff
@@ -84,9 +142,68 @@ def load_blueprints_loaders_from_module_path(module_path: str) -> Sequence[YamlB
...
+        if click.confirm(
+            "The redhat.vscode-yaml extension is not installed. Would you like to install it now?"
+        ):
+            run_vscode_cli_command(["--install-extension", "redhat.vscode-yaml"])
+
+
+def install_yaml_schema_extension(
+    dot_vscode_path: Path, yaml_dirs: List[Path], schema_paths: List[Path]
+) -> None:
+    """Builds a VS Code extension which provides the Red Hat YAML extension with the JSON schema files which
```
Comment (sryza) [inline examples/experimental/dagster-blueprints/dagster_blueprints/cli.py:174]: Nitpick: I found this wording confusing on the first couple reads; thought it meant that we build a VS Code extension that wraps the Red Had YAML extension.

File: examples/experimental/dagster-blueprints/dagster_blueprints/cli.py [hunk 5]
```diff
@@ -84,9 +142,68 @@ def load_blueprints_loaders_from_module_path(module_path: str) -> Sequence[YamlB
...
+        click.echo("redhat.vscode-yaml extension is already installed.")
+    else:
+        if click.confirm(
+            "The redhat.vscode-yaml extension is not installed. Would you like to install it now?"
+        ):
+            run_vscode_cli_command(["--install-extension", "redhat.vscode-yaml"])
+
+
+def install_yaml_schema_extension(
+    dot_vscode_path: Path, yaml_dirs: List[Path], schema_paths: List[Path]
```
Comment (sryza) [inline examples/experimental/dagster-blueprints/dagster_blueprints/cli.py:172]:
```suggestion
    dot_vscode_path: Path, yaml_dirs: Sequence[Path], schema_paths: Sequence[Path]
```

File: examples/experimental/dagster-blueprints/dagster_blueprints/cli.py [hunk 6]
```diff
@@ -84,9 +142,68 @@ def load_blueprints_loaders_from_module_path(module_path: str) -> Sequence[YamlB
...
+    """Builds a VS Code extension which provides the Red Hat YAML extension with the JSON schema files which
+    correspond to the YAML files in the provided directory.
+    """
+    extension_working_dir = dot_vscode_path / "dagster-blueprint-schema"
+    extension_package_json_path = extension_working_dir / "package.json"
+
+    template_package_json_path = Path(__file__).parent / "vscode_extension_package.json"
+    template_package_json = json.loads(template_package_json_path.read_text())
+    template_package_json["contributes"]["yamlValidation"] = [
+        {"fileMatch": f"{yaml_dir}/**/*.yaml", "url": f"{schema_path}"}
```
Comment (sryza) [inline examples/experimental/dagster-blueprints/dagster_blueprints/cli.py:183]: We should probably do '.yml' as well

File: examples/experimental/dagster-blueprints/dagster_blueprints/cli.py [hunk 7]
```diff
@@ -84,9 +142,68 @@ def load_blueprints_loaders_from_module_path(module_path: str) -> Sequence[YamlB
...
+    with zipfile.ZipFile(extension_zip_path, "w") as z:
+        z.write(extension_package_json_path, "extension/package.json")
+
+    click.echo(f"Packaged extension to {extension_zip_path}")
+
+    try:
+        run_vscode_cli_command(["--uninstall-extension", "dagster.dagster-blueprint-schema"])
+    except subprocess.CalledProcessError:
+        click.echo("No existing dagster.dagster-blueprint-schema extension to uninstall.")
+    run_vscode_cli_command(["--install-extension", os.fspath(extension_zip_path.resolve())])
```
Comment (sryza) [inline examples/experimental/dagster-blueprints/dagster_blueprints/cli.py:201]: Worth adding an echo to say that the extension was successfully installed

**Review by sryza — APPROVED** [review]

File: examples/experimental/dagster-blueprints/dagster_blueprints/cli.py (hunk 2 above)
Comment (benpankow) [reply examples/experimental/dagster-blueprints/dagster_blueprints/cli.py:110]: I think the right term is "workspace root folder" - there are multi-root workspaces, but a `.vscode` directory lives in that case in each root: 

https://code.visualstudio.com/docs/editor/multi-root-workspaces

> With multiple root folders in one workspace, it is possible to have a .vscode folder in each root folder defining the settings that should apply for that folder.

**Pull request merged**